### PR TITLE
transformations: New csl-stencil-materialize-stores pass

### PR DESCRIPTION
**Pull request by n-io**
Operates on memrefs and needs to be run after bufferization.

### CODE REVIEW

File: xdsl/transforms/csl_stencil_materialize_stores.py [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+                op.done_exchange.block.args[-1], [], Region(Block()), Region(Block())
+            ),
+            InsertPoint.at_start(op.done_exchange.block),
+        )
+
+        assert if_op.next_op, "Block cannot be empty"
+
+        if (
+            not isinstance(yld := op.done_exchange.block.last_op, csl_stencil.YieldOp)
+            or not len(yld.arguments) == 0
```
Comment (dk949) [inline xdsl/transforms/csl_stencil_materialize_stores.py:112]:
```suggestion
            or len(yld.arguments) != 0
```

File: xdsl/transforms/csl_stencil_materialize_stores.py [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+from xdsl.rewriter import InsertPoint
+from xdsl.utils.hints import isa
+
+
+@dataclass(frozen=True)
+class MaterializeInApplyDest(RewritePattern):
+    """
+    Lowers csl_stencil.yield to csl.return.
+    Note, the callbacks generated return no values, whereas the yield op
+    to be replaced may still report to yield values.
```
[the remark this answers is not in the document]
Comment (n-io) [reply xdsl/transforms/csl_stencil_materialize_stores.py:25]: Copy and pasta

File: xdsl/transforms/csl_stencil_materialize_stores.py [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+    def match_and_rewrite(self, op: csl_stencil.YieldOp, rewriter: PatternRewriter, /):
+        if not len(op.arguments) > 0:
+            return
+        assert isinstance(apply := op.parent_op(), csl_stencil.ApplyOp)
+
+        # the second callback stores yielded values to dest
+        if op.parent_region() == apply.done_exchange:
+            views: list[Operation] = []
+            add_args: list[SSAValue] = []
+            for src, dst in zip(op.arguments, apply.dest):
```
[the remark this answers is not in the document]
Comment (n-io) [reply xdsl/transforms/csl_stencil_materialize_stores.py:38]: Done

File: xdsl/transforms/csl_stencil_materialize_stores.py (hunk 1 above)
Comment (n-io) [reply xdsl/transforms/csl_stencil_materialize_stores.py:112]: Done

File: xdsl/transforms/csl_stencil_materialize_stores.py [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+    """
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: csl_stencil.YieldOp, rewriter: PatternRewriter, /):
+        if not len(op.arguments) > 0:
+            return
+        assert isinstance(apply := op.parent_op(), csl_stencil.ApplyOp)
+
+        # the second callback stores yielded values to dest
+        if op.parent_region() == apply.done_exchange:
```
[the remark this answers is not in the document]
Comment (n-io) [reply xdsl/transforms/csl_stencil_materialize_stores.py:35]: Done.

File: xdsl/transforms/csl_stencil_materialize_stores.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (n-io) [reply xdsl/transforms/csl_stencil_materialize_stores.py:25]: Fixed

File: xdsl/transforms/lower_csl_stencil.py [hunk 5]
```diff
@@ -198,31 +197,6 @@ class LowerYieldOp(RewritePattern):
```
[the remark this answers is not in the document]
Comment (n-io) [reply xdsl/transforms/lower_csl_stencil.py:198]: Updated, thanks

**Review by dk949 — APPROVED** [review]

**Pull request merged**